upgrade/v1beta1/upgrade.go

```diff
@@ -0,0 +1,183 @@
+package v1beta1
+
+import (
+	"context"
+
+	"github.com/jinzhu/copier"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	restclient "k8s.io/client-go/rest"
+
+	longhorn_v1beta1 "github.com/longhorn/longhorn-manager/upgrade/v1beta1/k8s/pkg/apis/longhorn/v1beta1"
+	lhclientset_v1beta1 "github.com/longhorn/longhorn-manager/upgrade/v1beta1/k8s/pkg/client/clientset/versioned"
+	"github.com/longhorn/longhorn-manager/upgrade/v1beta1/types"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1beta1 to v1beta2:"
+)
+
+func IsCRDVersionMatch(config *restclient.Config, namespace string) (bool, error) {
+	lhClientV1beta1, err := lhclientset_v1beta1.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhorn_v1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return false, errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	if _, err := lhClientV1beta1.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameDefaultEngineImage), metav1.GetOptions{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			// cannot find the setting,
+			logrus.Infof("Longhorn CRD API v1beta1 not found")
+			return false, nil
+		}
+		return false, errors.Wrap(err, "unable to verify if version matches v1beta1")
+	}
+	logrus.Infof("Detected Longhorn CRD API v1beta1")
+	return true, nil
+}
+
+func UpgradeFromV1beta1ToV1beta2(config *restclient.Config, namespace string, lhClient *lhclientset.Clientset) error {
+	lhClientV1beta1, err := lhclientset_v1beta1.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhorn_v1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	volumes, err := lhClientV1beta1.LonghornV1beta1().Volumes(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "upgrade: v1beta1: unable to list volumes")
+	}
+	for _, old := range volumes.Items {
+		new := &longhorn.Volume{}
+
+		if err := copier.Copy(&new, &old); err != nil {
```

Comment:
Will there be an error since the field `condition` is different between old volume definition and the new one?

Reply:
No, copier.Copy can copy to different struct and will skip the fields having different datatype.